File: dashlane-project/Dashlane/src/main/java/com/dashlane/ui/activities/HomeActivity.java
package com.dashlane.ui.activities;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.Window;

import com.dashlane.R;
import com.dashlane.abtesting.RemoteAbTestManager;
import com.dashlane.analytics.metrics.time.SpentTimeOnViewManager;
import com.dashlane.announcements.modules.KeyboardAutofillAnnouncementModule;
import com.dashlane.braze.BrazeWrapper;
import com.dashlane.core.DataSync;
import com.dashlane.dagger.singleton.SingletonProvider;
import com.dashlane.debug.DaDaDa;
import com.dashlane.events.AppEvents;
import com.dashlane.events.PremiumStatusChangedEvent;
import com.dashlane.login.lock.LockManager;
import com.dashlane.navigation.NavigationConstants;
import com.dashlane.navigation.NavigationUtils;
import com.dashlane.session.Session;
import com.dashlane.ui.activities.fragments.vault.HiddenImpala;
import com.dashlane.ui.dialogs.fragment.SpaceRevokedDialog;
import com.dashlane.ui.dialogs.fragments.NotificationDialogFragment;
import com.dashlane.ui.menu.DashlaneMenuView;
import com.dashlane.ui.util.ActionBarUtil;
import com.dashlane.ui.util.DrawerLayoutUtilsKt;
import com.dashlane.ui.widgets.view.MainDrawerToggle;
import com.dashlane.util.AppShortcutsUtil;
import com.dashlane.util.DeviceUtils;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.lifecycle.LifecycleOwnerKt;
import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class HomeActivity extends DashlaneActivity
        implements ActionBarUtil.DrawerLayoutProvider, MenuContainer {

    private static final String SAVED_STATE_HAS_CHANGED_CONFIGURATIONS = "saved_state_has_changed_configurations";

    @NonNull
    protected DrawerLayout mNavigationDrawer;
    private MainDrawerToggle mDrawerToggle;
    private DashlaneMenuView mMenuFrame;
    private boolean mIsResume;
    private LockManager mLockManager;

    @Inject
    RemoteAbTestManager remoteAbTestManager;

    @Inject
    AppShortcutsUtil appShortcutsUtil;

    @Inject
    DataSync mDataSync;

    @Inject
    BrazeWrapper brazeWrapper;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        configureAndroidActivityFeatures();
        super.onCreate(savedInstanceState);
        Session session = SingletonProvider.getSessionManager().getSession();
        if (session == null) {
            
            NavigationUtils.logoutAndCallLoginScreen(this, false);
            return;
        }

        mLockManager = SingletonProvider.getComponent().getLockRepository()
                .getLockManager(session);

        setContentView(R.layout.activity_home_activity_layout);
        mNavigationDrawer = findViewById(R.id.drawer_layout);
        mMenuFrame = findViewById(R.id.menu_frame);

        initializeNavigationDrawer();

        
        HiddenImpala.configureForHomeActivity(this);
    }

    @Override
    public void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        handleDeepLink(getIntent());
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        if (mDrawerToggle != null) {
            mDrawerToggle.onConfigurationChanged(newConfig);
        }
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        handleDeepLink(intent);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        
        
        if (mDrawerToggle.onOptionsItemSelected(item)) {
            return true;
        }
        

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onSupportNavigateUp() {
        return SingletonProvider.getNavigator().navigateUp();
    }

    private void forceCloseNavigationDrawer(int gravity) {
        mNavigationDrawer.closeDrawer(gravity);
    }

    @Override
    public boolean isNavigationDrawerVisible() {
        return mNavigationDrawer.isDrawerOpen(GravityCompat.START);
    }

    @Override
    public void disableMenuAccess(boolean bool) {
        if (bool) {
            mNavigationDrawer.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED, GravityCompat.START);
        } else {
            mNavigationDrawer.setDrawerLockMode(DrawerLayout.LOCK_MODE_UNLOCKED, GravityCompat.START);
        }
    }

    @Override
    public void onBackPressed() {
        if (isNavigationDrawerVisible()) {
            forceCloseNavigationDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public DrawerLayout getNavigationDrawer() {
        return mNavigationDrawer;
    }

    private void onPremiumStatusChanged(PremiumStatusChangedEvent event) {
        SingletonProvider.getAppEvents().clearLastEvent(PremiumStatusChangedEvent.class);
        if (event.premiumPlanChanged()) {
            SingletonProvider.getAccessibleOffersCache().flushCache();
            loadAccessibleOffers();
            mMenuFrame.refresh();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();

        SingletonProvider.getUserFeature().launchRefreshIfNeeded();

        remoteAbTestManager.launchRefreshIfNeeded();

        
        SingletonProvider.getOfflineExperimentsReporter().launchReportIfNeeded();

        SingletonProvider.getBillingVerificator().verifyAndConsumePurchaseIfNeeded();

        final DaDaDa daDaDa = SingletonProvider.getDaDaDa();
        if (daDaDa.isEnabled() && daDaDa.hasDeepLink()) {
            String link = daDaDa.getDeepLink();
            NotificationDialogFragment.Builder builder = new NotificationDialogFragment.Builder();
            builder.setNegativeButtonText("Cancel")
                   .setPositiveButtonDeepLink(link)
                   .setPositiveButtonText("Go")
                   .setMessage("Go to $link");
            builder.build().show(getSupportFragmentManager(), link);
        }

        SingletonProvider.getSecurityHelper().showPopupPinCodeDisableIfWasEnable(this);

        SingletonProvider.getComponent().getDeviceUpdateManager().updateIfNeeded();

        KeyboardAutofillAnnouncementModule.Companion
                .setRequireDisplayIfNeeded(SingletonProvider.getUserPreferencesManager(), getIntent());
    }

    @Override
    protected void onResume() {
        super.onResume();
        brazeWrapper.openSession(this);

        
        if (isFinishing()) {
            mIsResume = false;
            return;
        }
        mIsResume = true;
        if (isNavigationDrawerVisible()) {
            DeviceUtils.hideKeyboard(mNavigationDrawer);
        }
        final AppEvents appEvents = SingletonProvider.getAppEvents();
        appEvents.register(this, PremiumStatusChangedEvent.class, true, premiumStatusChangedEvent -> {
            onPremiumStatusChanged(premiumStatusChangedEvent);
            return null;
        });

        if (!mLockManager.isLocked()) {
            SpentTimeOnViewManager.getInstance().enterView(getLocalClassName());
        }

        SpaceRevokedDialog.showIfNecessary(this, getSupportFragmentManager());
        SpaceRevokedDialog.listenUpcoming(this);
        mMenuFrame.refresh();
        loadAccessibleOffers();
        mDataSync.maySync();
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(SAVED_STATE_HAS_CHANGED_CONFIGURATIONS, isChangingConfigurations());
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mIsResume) {
            
            AppEvents appEvents = SingletonProvider.getAppEvents();
            appEvents.unregister(this, PremiumStatusChangedEvent.class);
            if (SingletonProvider.getSessionManager().getSession() != null) {
                SpentTimeOnViewManager.getInstance().leaveView(getLocalClassName());
            }
        }
        mIsResume = false;
        SpaceRevokedDialog.stopUpcomingListener();
    }

    @Override
    protected void onStop() {
        super.onStop();
        
        appShortcutsUtil.refreshShortcuts(this);
    }

    public boolean isResume() {
        return mIsResume;
    }

    private void initializeNavigationDrawer() {
        ActionBarUtil actionBarUtil = getActionBarUtil();
        actionBarUtil.setup();
        mDrawerToggle =
                new MainDrawerToggle(this, mNavigationDrawer, actionBarUtil.getToolbar());
        DrawerLayoutUtilsKt.setup(mNavigationDrawer, mDrawerToggle, actionBarUtil, mMenuFrame);
        SingletonProvider
                .getNavigator()
                .addOnDestinationChangedListener(
                        (controller, destination, arguments) -> forceCloseNavigationDrawer(GravityCompat.START)
                );
    }

    private void configureAndroidActivityFeatures() {
        supportRequestWindowFeature(Window.FEATURE_INDETERMINATE_PROGRESS);
    }

    private void loadAccessibleOffers() {
        SingletonProvider.getAccessibleOffersCache()
                .prefetchProductsForCurrentUser(LifecycleOwnerKt.getLifecycleScope(this));
    }

    private void handleDeepLink(Intent intent) {
        if (intent != null && intent.hasExtra(NavigationConstants.STARTED_WITH_INTENT)) {
            Intent startIntent = intent.getParcelableExtra(NavigationConstants.STARTED_WITH_INTENT);
            if (startIntent != null) {
                SingletonProvider.getNavigator().handleDeepLink(startIntent);
                intent.removeExtra(NavigationConstants.STARTED_WITH_INTENT);
                setIntent(intent);
            }
        }
    }
}

File: dashlane-project/Dashlane/src/main/java/com/dashlane/ui/activities/HomeActivity.kt
package com.dashlane.ui.activities

import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.activity.OnBackPressedCallback
import androidx.activity.viewModels
import androidx.core.view.GravityCompat
import androidx.core.view.WindowCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.dashlane.R
import com.dashlane.abtesting.OfflineExperimentReporter
import com.dashlane.abtesting.RemoteAbTestManager
import com.dashlane.accountstatus.AccountStatusRepository
import com.dashlane.accountstatus.premiumstatus.planName
import com.dashlane.announcements.AnnouncementCenter
import com.dashlane.announcements.modules.KeyboardAutofillAnnouncementModule
import com.dashlane.braze.BrazeWrapper
import com.dashlane.databinding.ActivityHomeActivityLayoutBinding
import com.dashlane.debug.services.DaDaDaLogin
import com.dashlane.device.DeviceUpdateManager
import com.dashlane.events.AppEvents
import com.dashlane.featureflipping.FeatureFlipManager
import com.dashlane.hardwaresecurity.SecurityHelper
import com.dashlane.lock.LockManager
import com.dashlane.lock.LockType
import com.dashlane.navigation.NavigationConstants
import com.dashlane.navigation.Navigator
import com.dashlane.preference.PreferencesManager
import com.dashlane.premium.StoreOffersCache
import com.dashlane.session.SessionManager
import com.dashlane.session.authorization
import com.dashlane.sync.DataSync
import com.dashlane.ui.activities.fragments.vault.HiddenImpala
import com.dashlane.ui.dialogs.fragment.TeamRevokedDialogDisplayer
import com.dashlane.ui.dialogs.fragments.NotificationDialogFragment
import com.dashlane.ui.menu.MenuViewModel
import com.dashlane.ui.menu.menuScreen
import com.dashlane.ui.premium.inappbilling.BillingVerification
import com.dashlane.ui.util.ActionBarUtil.DrawerLayoutProvider
import com.dashlane.ui.util.setup
import com.dashlane.ui.widgets.view.MainDrawerToggle
import com.dashlane.util.AppShortcutsUtil
import com.dashlane.util.DeviceUtils.hideKeyboard
import com.dashlane.util.getParcelableExtraCompat
import dagger.hilt.android.AndroidEntryPoint
import java.lang.ref.WeakReference
import javax.inject.Inject
import kotlinx.coroutines.launch

@AndroidEntryPoint
class HomeActivity : DashlaneActivity(), DrawerLayoutProvider, MenuContainer {
    @Inject
    lateinit var announcementCenter: AnnouncementCenter

    @Inject
    lateinit var remoteAbTestManager: RemoteAbTestManager

    @Inject
    lateinit var appShortcutsUtil: AppShortcutsUtil

    @Inject
    lateinit var mDataSync: DataSync

    @Inject
    lateinit var brazeWrapper: BrazeWrapper

    @Inject
    lateinit var offlineExperimentReporter: OfflineExperimentReporter

    @Inject
    lateinit var featureFlipManager: FeatureFlipManager

    @Inject
    lateinit var billingVerification: BillingVerification

    @Inject
    lateinit var storeOffersCache: StoreOffersCache

    @Inject
    lateinit var appEvents: AppEvents

    @Inject
    lateinit var securityHelper: SecurityHelper

    @Inject
    lateinit var sessionManager: SessionManager

    @Inject
    lateinit var lockManager: LockManager

    @Inject
    lateinit var dadadaLogin: DaDaDaLogin

    @Inject
    lateinit var deviceUpdateManager: DeviceUpdateManager

    @Inject
    lateinit var preferencesManager: PreferencesManager

    @Inject
    lateinit var teamSpaceRevokedDialogDisplayer: TeamRevokedDialogDisplayer

    @Inject
    lateinit var accountStatusRepository: AccountStatusRepository

    @Inject
    lateinit var navigator: Navigator

    private lateinit var navigationDrawer: DrawerLayout
    private val menuViewModel: MenuViewModel by viewModels()
    private var drawerToggle: MainDrawerToggle? = null
    var isResume = false
        private set

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val session = sessionManager.session
        if (session == null) {
            
            navigator.logoutAndCallLoginScreen(this)
            return
        }
        val homeViewBinding = ActivityHomeActivityLayoutBinding.inflate(layoutInflater)
        setContentView(homeViewBinding.root)
        navigationDrawer = homeViewBinding.drawerLayout
        menuScreen(homeViewBinding, menuViewModel)
        navigator.addOnDestinationChangedListener(menuViewModel)
        initializeNavigationDrawer()

        
        val drawerCallback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (isNavigationDrawerVisible()) {
                    forceCloseNavigationDrawer()
                }
            }
        }
        navigationDrawer.addDrawerListener(object : DrawerLayout.SimpleDrawerListener() {
            override fun onDrawerClosed(drawerView: View) {
                drawerCallback.remove()
            }

            override fun onDrawerOpened(drawerView: View) {
                onBackPressedDispatcher.addCallback(this@HomeActivity, drawerCallback)
            }
        })

        lifecycleScope.launch {
            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
                accountStatusRepository.accountStatusState.collect { accountStatusState ->
                    accountStatusState[sessionManager.session]?.let {
                        onPremiumStatusChanged(it.premiumStatus.planName)
                        teamSpaceRevokedDialogDisplayer.onStatusChanged(WeakReference(this@HomeActivity))
                    }
                }
            }
        }

        
        HiddenImpala.configureForHomeActivity(this)
    }

    public override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        handleDeepLink(intent)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        drawerToggle?.onConfigurationChanged(newConfig)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleDeepLink(intent)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        
        
        return if (drawerToggle?.onOptionsItemSelected(item) == true) {
            true
        } else {
            
            super.onOptionsItemSelected(item)
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        return navigator.navigateUp()
    }

    private fun forceCloseNavigationDrawer() {
        navigationDrawer.closeDrawer(GravityCompat.START)
    }

    override fun isNavigationDrawerVisible(): Boolean {
        return navigationDrawer.isDrawerOpen(GravityCompat.START)
    }

    override fun disableMenuAccess(bool: Boolean) {
        if (bool) {
            navigationDrawer.setDrawerLockMode(
                DrawerLayout.LOCK_MODE_LOCKED_CLOSED,
                GravityCompat.START
            )
        } else {
            navigationDrawer.setDrawerLockMode(DrawerLayout.LOCK_MODE_UNLOCKED, GravityCompat.START)
        }
    }

    override fun getNavigationDrawer(): DrawerLayout {
        return navigationDrawer
    }

    private fun onPremiumStatusChanged(newSubscription: String?) {
        storeOffersCache.flushCacheIfSubscriptionHasChanged(newSubscription)
        loadAccessibleOffers()
        menuViewModel.refresh()
    }

    override fun onStart() {
        super.onStart()
        sessionManager.session?.authorization?.let { authorization ->
            featureFlipManager.launchRefreshIfNeeded(authorization)
            remoteAbTestManager.launchRefreshIfNeeded(authorization)
            offlineExperimentReporter.launchReportIfNeeded(authorization)
        }

        
        billingVerification.verifyAndConsumePurchaseIfNeeded()
        if (dadadaLogin.isEnabled && dadadaLogin.hasDeepLink()) {
            val link = dadadaLogin.deepLink
            val builder = NotificationDialogFragment.Builder()
            builder.setNegativeButtonText("Cancel")
                .setPositiveButtonDeepLink(link)
                .setPositiveButtonText("Go")
                .setMessage("Go to \$link")
            builder.build().show(supportFragmentManager, link)
        }

        
        val username = sessionManager.session?.username
        if (username != null && !securityHelper.isDeviceSecured(username)) {
            val locks = lockManager.getLocks(username)

            if (LockType.PinCode in locks || LockType.Biometric in locks) {
                lockManager.removeLock(username, LockType.PinCode)
                lockManager.removeLock(username, LockType.Biometric)
                securityHelper.showPopupPinCodeDisable(this)
            }
        }

        deviceUpdateManager.updateIfNeeded()
        KeyboardAutofillAnnouncementModule.setRequireDisplayIfNeeded(preferencesManager[username], intent)
        announcementCenter.start(this)
    }

    override fun onResume() {
        super.onResume()
        brazeWrapper.openSession(this)

        
        if (isFinishing) {
            isResume = false
            return
        }
        isResume = true
        if (isNavigationDrawerVisible()) {
            hideKeyboard(navigationDrawer)
        }
        menuViewModel.refresh()
        loadAccessibleOffers()
        mDataSync.maySync()
    }

    override fun onPause() {
        super.onPause()
        isResume = false
    }

    override fun onStop() {
        super.onStop()
        
        appShortcutsUtil.refreshShortcuts(this)
        announcementCenter.stop(this)
    }

    private fun initializeNavigationDrawer() {
        actionBarUtil.setup()
        drawerToggle = MainDrawerToggle(this, navigationDrawer, actionBarUtil.toolbar)
        navigationDrawer.setup(drawerToggle!!, actionBarUtil, findViewById(R.id.menu_frame))
        navigator.addOnDestinationChangedListener { _, _, _ ->
            forceCloseNavigationDrawer()
        }
    }

    private fun loadAccessibleOffers() {
        storeOffersCache.prefetchProductsForCurrentUser(this.lifecycleScope)
    }

    private fun handleDeepLink(intent: Intent?) {
        if (intent != null && intent.hasExtra(NavigationConstants.STARTED_WITH_INTENT)) {
            val startIntent = intent.getParcelableExtraCompat<Intent>(NavigationConstants.STARTED_WITH_INTENT)
            if (startIntent != null) {
                navigator.handleDeepLink(startIntent)
                intent.removeExtra(NavigationConstants.STARTED_WITH_INTENT)
                setIntent(intent)
            }
        }
    }
}